File: bench/src/spectralnorm/spectralnorm.py
import sys
from math import sqrt
def eval_A(i, j):
 return 1.0 / ((i + j) * (i + j + 1) // 2 + i + 1)
def multiply_Av(v, n):
 return [sum(v[j] * eval_A(i, j) for j in range(n)) for i in range(n)]
def multiply_Atv(v, n):
 return [sum(v[j] * eval_A(j, i) for j in range(n)) for i in range(n)]
def multiply_AtAv(v, n):
 return multiply_Atv(multiply_Av(v, n), n)
def main():
 n = int(sys.argv[1]) if len(sys.argv) > 1 else 5500
 u = [1.0] * n
 v = [0.0] * n
 for _ in range(10):
  v = multiply_AtAv(u, n)
  u = multiply_AtAv(v, n)
 vBv = sum(u[i] * v[i] for i in range(n))
 vv = sum(v[i] * v[i] for i in range(n))
 sys.stdout.write("%.9f\n" % sqrt(vBv / vv))
main()

File: bench/src/spectralnorm/spectralnorm.js
function eval_A(i, j) {
return 1.0 / ((i + j) * (i + j + 1) / 2 + i + 1);
}
function multiply_Av(v, n) {
const out = new Float64Array(n);
for (let i = 0; i < n; i++) {
let s = 0.0;
for (let j = 0; j < n; j++) {
s += v[j] * eval_A(i, j);
}
out[i] = s;
}
return out;
}
function multiply_Atv(v, n) {
const out = new Float64Array(n);
for (let i = 0; i < n; i++) {
let s = 0.0;
for (let j = 0; j < n; j++) {
s += v[j] * eval_A(j, i);
}
out[i] = s;
}
return out;
}
function multiply_AtAv(v, n) {
return multiply_Atv(multiply_Av(v, n), n);
}
function main() {
const n = process.argv[2] ? parseInt(process.argv[2], 10) : 5500;
let u = new Float64Array(n).fill(1.0);
let v = new Float64Array(n);
for (let k = 0; k < 10; k++) {
v = multiply_AtAv(u, n);
u = multiply_AtAv(v, n);
}
let vbv = 0.0;
let vv = 0.0;
for (let i = 0; i < n; i++) {
vbv += u[i] * v[i];
vv += v[i] * v[i];
}
process.stdout.write((Math.sqrt(vbv / vv)).toFixed(9) + "\n");
}
main();
